titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/OpportunisticCpuSchedulingTest.java

```diff
@@ -0,0 +1,196 @@
+        jobsScenarioBuilder.takeJob(0).template(ScenarioTemplates.killJob());
+
+        // opportunistic CPUs are available again, and window is still valid (6h expiration)
+        jobsScenarioBuilder.schedule(jobDescriptor, jobScenarioBuilder -> jobScenarioBuilder
+                .template(ScenarioTemplates.launchJob())
+                .allTasks(taskScenarioBuilder -> taskScenarioBuilder
+                        .expectTaskOnAgent()
+                        .expectTaskContext(TASK_ATTRIBUTES_AGENT_ID, opportunisticCpus.agentId)
+                        .expectTaskContext(TASK_ATTRIBUTES_OPPORTUNISTIC_CPU_ALLOCATION, opportunisticCpus.allocationId)
+                        .expectTaskContext(TASK_ATTRIBUTES_OPPORTUNISTIC_CPU_COUNT, "4")
+                ));
+    }
+
+    @Test(timeout = TEST_TIMEOUT_MS)
+    public void expiredOpportunisticCpuAvailability() throws Exception {
+        addOpportunisticCpusToFlex(10, Instant.now().minus(Duration.ofMillis(1)));
+
+        JobDescriptor<BatchJobExt> jobDescriptor = BATCH_JOB_WITH_RUNTIME_PREDICTION.but(j ->
+                j.getContainer().but(c -> c.getContainerResources().toBuilder().withCpu(4))
+        );
+        jobsScenarioBuilder.schedule(jobDescriptor, jobScenarioBuilder -> jobScenarioBuilder
+                .template(ScenarioTemplates.launchJob())
+                .allTasks(taskScenarioBuilder -> taskScenarioBuilder
+                        .expectTaskOnAgent()
+                        .assertTask(task -> !task.getTaskContext().containsKey(TASK_ATTRIBUTES_OPPORTUNISTIC_CPU_ALLOCATION) &&
+                                        !task.getTaskContext().containsKey(TASK_ATTRIBUTES_OPPORTUNISTIC_CPU_COUNT),
+                                "Not scheduled on opportunistic CPUs")
+                )
+        );
+    }
+
+    @Test(timeout = TEST_TIMEOUT_MS)
+    public void someOpportunisticCpusAvailable() throws Exception {
+        OpportunisticCpus opportunisticCpus = addOpportunisticCpusToFlex(2, Instant.now().plus(Duration.ofHours(6)));
+
+        JobDescriptor<BatchJobExt> jobDescriptor = BATCH_JOB_WITH_RUNTIME_PREDICTION.but(j ->
+                j.getContainer().but(c -> c.getContainerResources().toBuilder().withCpu(4))
+        );
+
+        jobsScenarioBuilder.schedule(jobDescriptor, jobScenarioBuilder -> jobScenarioBuilder
+                .template(ScenarioTemplates.launchJob())
+                .allTasks(taskScenarioBuilder -> taskScenarioBuilder
+                        .expectTaskOnAgent()
+                        .expectTaskContext(TASK_ATTRIBUTES_AGENT_ID, opportunisticCpus.agentId)
+                        .expectTaskContext(TASK_ATTRIBUTES_OPPORTUNISTIC_CPU_ALLOCATION, opportunisticCpus.allocationId)
+                        .expectTaskContext(TASK_ATTRIBUTES_OPPORTUNISTIC_CPU_COUNT, "2")
+                )
+        );
+    }
+
+    private OpportunisticCpus addOpportunisticCpusToFlex(int count, Instant expiresAt) {
```

Comment:
This could be added to `InstanceGroupsScenarioBuilder`

Reply:
Done.